src/PathOrderOptimizer.h

```diff
@@ -179,18 +179,25 @@ class PathOrderOptimizer
      * This reorders the \ref paths field and fills their starting vertices and
      * directions.
      */
-    void optimize()
+    void optimize(bool mirrored = false)
     {
         if(paths.empty())
         {
             return;
         }
 
+        const std::function<Point(const Point&)> maybe_mirror =
+            [&mirrored](const Point& p)
+            {
+                return Point{ p.X, p.Y * (mirrored ? -1 : 1) };
+            };
```

Comment:
I'm not entirely sure what this function aims to do, if I'm not correct let me know. The aim of this mirror function is to not only reverse the order of the polygon itself, but also reverse the order of the polygons itself. I'm guessing this is specifically needed for infill as the print direction of these extrusions should also alternate.

I'm not sure if this solves the issue though. Conciser the image attached. The tool paths are traversed as if they were in the mirrored position, however using the "travel to closest endpoint" path optimizing strategy does not result in an reversed order of tool paths for this case; in fact, the order is not changed at all.
![Untitled Diagram](https://user-images.githubusercontent.com/6638028/145837678-2364d8e9-f895-4d6d-9e50-a27582a3731a.png)

May I suggest two alternative approach
1. Instead of mirroring move the point to the _opposite location of the build plate_. So `point(X, Y) -> point(build_plate_width - X, build_plate_height - Y)`. Problem with this solution is that this wouldn't work as well if the starting location is close to the center of the build plate. (also as the distances between mirrored points is the same as before mirroring it is semantically the same to _only_ mirror the starting position for alternating layers)
2. Calculate the optimize path order normally (without mirror), and if the tool paths should be reversed reverse the order of the polygons in the optimized order, and for all open polygons reverse the order of the points. The initial travel moves will be a bit illogical as a result. but that is a compromise.

Reply:
I'm not really sure I see a reason to reverse the order of open polylines at all. The infill is not expected to change direction, is it?

Regardless of in which direction the polygons are printed, you'd start and end on the same spot, so for the travel length optimisation it shouldn't matter.

So I'm inclined to go with your second option but also simplify it to not adjust open polylines any more.

Reply:
Great analysis.

I think this can be removed if it gives us cause to think this deeply about it: It was just something extra that isn't in the acceptance criteria; propose to remove.

Reply:
This is changed in 8368fc28a90fb457202be19483c2f3799bd40b6c. The new solution is hopefully simpler, although it does involve another copy of the `Path` instance.

This turns out to have been the cause of the misplaced lines as well. With polylines, the direction of the line is calculated based on the optimal starting vertex. This was reversed as well, causing the direction of polylines to be the wrong way around again. In the new implementation everything is reversed exactly once so the extra lines are gone.